quiche/src/lib.rs

```diff
@@ -3390,30 +3388,48 @@ impl<F: BufFactory> Connection<F> {
             });
         }
 
-        // Following flag used to upgrade datagram size, if probe is successful.
-        let mut successful_pmtud_probe = false;
-
         // Process acked frames. Note that several packets from several paths
         // might have been acked by the received packet.
         for (_, p) in self.paths.iter_mut() {
             for acked in p.recovery.get_acked_frames(epoch) {
                 match acked {
                     frame::Frame::Ping {
                         mtu_probe: Some(mtu_probe),
-                    } => {
-                        let current_mtu = p.pmtud.get_current_mtu();
-                        p.pmtud.set_current_mtu(cmp::max(current_mtu, mtu_probe));
+                    } =>
+                        if let Some(pmtud) = p.pmtud.as_mut() {
+                            trace!(
+                                "{} pmtud probe acked; probe size {:?}",
+                                self.trace_id,
+                                mtu_probe
+                            );
 
-                        // Stop sending path MTU probes after successful probe.
-                        p.pmtud.set_should_probe(false);
-                        successful_pmtud_probe = true;
+                            pmtud.successful_probe(mtu_probe);
+
+                            qlog_with_type!(
+                                EventType::ConnectivityEventType(
+                                    ConnectivityEventType::MtuUpdated
+                                ),
+                                self.qlog,
+                                q,
+                                {
+                                    let pmtu_data = EventData::MtuUpdated(
+                                        qlog::events::connectivity::MtuUpdated {
+                                            old: Some(
+                                                p.recovery.max_datagram_size()
+                                                    as u16,
+                                            ),
+                                            new: mtu_probe as u16,
+                                            done: Some(true),
+                                        },
+                                    );
+
+                                    q.add_event_data_with_instant(pmtu_data, now)
+                                        .ok();
+                                }
+                            );
 
-                        trace!(
-                            "{} pmtud acked; current mtu: {:?}",
-                            self.trace_id,
-                            p.pmtud.get_current_mtu()
-                        );
-                    },
+                            p.recovery.pmtud_update_max_datagram_size(mtu_probe);
```

Comment:
`successful_probe()` ends up respecting `std::cmp::max(mtu_probe, self.largest_successful_probe_size)`, but we always call `p.recovery.pmtud_update_max_datagram_size()` with `mtu_probe`. I think this is meant to be resilient against out-of-order packets / retransmissions in ACKs from the peer, and in those cases we may inadvertently lower MTU